jaxopt/_src/osqp.py

```diff
@@ -0,0 +1,602 @@
+
+    certif_dual_infeasible = jnp.logical_and(jnp.logical_and(certif_Q <= criterion, certif_c <= criterion), certif_A)
+
+    if self.verbose >= 2:
+      print(f"certif_Q={certif_Q} certif_c={certif_c} certif_A={certif_A} criterion={criterion}, Adx={Adx}, certif_l={certif_l}, certif_u={certif_u}")
+
+    # infeasible dual implies either infeasible primal, either unbounded primal.
+    return jax.lax.cond(certif_dual_infeasible,
+      lambda _: (0., OSQP.DUAL_INFEASIBLE),  # dual unfeasible; exit the main loop with error = 0.
+      lambda _: (error, status),
+      operand=None)
+
+  def _check_primal_infeasability(self, error, status, delta_y, ATdy, l, u):
+    criterion = self.primal_infeasible_tol * tree_inf_norm(delta_y)
+    certif_A  = tree_inf_norm(ATdy)
+    bounded_l = tree_where(tree_map(lambda li: li == -jnp.inf, l), 0., l)  # replace inf bounds by zero
+    bounded_u = tree_where(tree_map(lambda ui: ui == jnp.inf, u), 0., u)
+    dy_plus   = tree_map(jax.nn.relu, delta_y)
+    dy_minus  = tree_negative(tree_map(jax.nn.relu, tree_negative(delta_y)))
+    certif_lu = tree_add(tree_vdot(bounded_l, dy_minus), tree_vdot(bounded_u, dy_plus))
+    certif_primal_infeasible = jnp.logical_and(certif_A  <= criterion, certif_lu  <= criterion)
+
+    if self.verbose >= 2:
+      print(f"certif_A={certif_A}, certif_lu={certif_lu}, criterion={criterion}")
+
+    return jax.lax.cond(certif_primal_infeasible,
+      lambda _: (0.,  # primal unfeasible; exit the main loop with error = 0.
+                OSQP.PRIMAL_INFEASIBLE),  
+      lambda _: (error, status),  # primal feasible or unbounded (depends of dual feasability).
+      operand=None)  
+
+  def _check_infeasability(self, prev_sol, sol, error, status, Q, c, A, l, u):
+    delta_x = tree_sub(sol.primal[0], prev_sol.primal[0])
+    delta_y = tree_sub(sol.dual_eq, prev_sol.dual_eq)
+    Adx, ATdy = A.matvec_and_rmatvec(delta_x, delta_y)
+
+    error, status = self._check_dual_infeasability(error, status, delta_x, Q, c, Adx, l, u)
+    error, status = self._check_primal_infeasability(error, status, delta_y, ATdy, l, u)
+
+    jit, _ = self._get_loop_options()
+    if not jit:
+      if status == OSQP.PRIMAL_INFEASIBLE:
+        raise ValueError(f"Primal infeasible. Certificate: y(t+1)-y(t) = {delta_y}.")
+      if status == OSQP.DUAL_INFEASIBLE:
+        raise ValueError(f"Dual infeasible. Certificate: x(t+1)-x(t) = {delta_x}.")
+
+    return error, status
+
+  def _check_termination_conditions(self, primal_residuals, dual_residuals,
+                                    old_params, params, Q, c, A, l, u):
+    error, status = self._compute_error(primal_residuals, dual_residuals)
+    if self.check_primal_dual_infeasability:
+      error, status = self._check_infeasability(old_params, params, error, status, Q, c, A, l, u)
+    return error, status
+
+  def _solve_linear_system(self, params, Q, c, A, rho_bar, state):
```

Comment:
Not necessary to do it in this PR but eventually, we may want to factor out this portion of the code, i.e., create a class specialized for equality-constrained QPs, that would be used in both `QuadraticProgramming` (when no inequality constraint is provided) and in `OSQP`.

Reply:
Indeed ! Note that because of `rho_bar`, `self.sigma` and dual variable `y` we are solving a _regularized_ version of the equality constrained QP. This ensures the problem is well posed but I am not sure the optima will coincide if `y` does not have the correct value.